### PR TITLE
add kube-env variable to block traffic to metadataserver

### PR DESCRIPTION
**Pull request by mikedanese**
@Q-Lee @thockin

### CODE REVIEW

**Comment by cjcullen** [comment]
This LGTM as a replacement for #42010. If we want to do smarter filtering, we can enable that flag, and then run a host network daemonset that inserts a DNAT rule in that chain. I'd like to hear @thockin's thoughts on whether that sounds reasonable.

**Comment by mikedanese** [comment]
> we can enable that flag, and then run a host network daemonset that inserts a DNAT rule in that chain

We would -t filter -I KUBE-METADATASERVER -j ACCEPT then add a DNAT in the nat table like we do in #42010

File: cluster/gce/configure-vm.sh [hunk 1]
```diff
@@ -89,6 +89,18 @@ ensure-local-disks() {
   done
 }
 
+function config-ip-firewall {
+  echo "Configuring IP firewall rules"
+
+  iptables -N KUBE-METADATA-SERVER
+  iptables -A FORWARD -p tcp -d 169.254.169.254 --dport 80 -j KUBE-METADATA-SERVER
+  iptables -A KUBE-METADATA-SERVER -j ACCEPT
```
Comment (mikedanese) [inline cluster/gce/configure-vm.sh]: I think this rule is unnecessary.

**Comment by thockin** [comment]
LGTM

**Comment by justinsb** [comment]
Two questions:

1) Does this apply to host-network pods?  i.e. can we use this to give k-c-m a different role from apiserver.
2) Given #1, how can we give the IAM-agent a special permission?

**Comment by mikedanese** [comment]
1) No.
2) you would have to build a container aware proxy.

**Comment by thockin** [comment]
Is that what we want?  We can refine this rule to only capture pods

**Comment by justinsb** [comment]
Yes, for the container-aware proxy there are a few solutions, and kube2iam is widely used on kuberentes with AWS

The challenge I think is figuring out iptables rules (or similar) that:

* don't let hostNetwork pods bypass it
* let us give different permissions to k-c-m vs apiserver
* still allow us to run the container proxy kubernetes-managed, while letting it reach the IAM service and not blocking bootstrapping

I'm not sure it is possible to achieve all of these; but we should decide what we are willing to sacrifice.

On the one hand this PR is progress (and I like the approach), on the other hand it could turn out to be moving in the wrong direction.  Though I can't see a world in which we _don't_ have this rule for hostNetwork=false pods.

**Comment by mikedanese** [comment]
> Is that what we want? We can refine this rule to only capture pods

Can you clarify?

> don't let hostNetwork pods bypass it

Host network pods should be considered highly privileged. Rather then fixing this, we should recommend that users don't use them. Admins can enforce this with pod security policies.

What we want for 1.6 is to secure the unprivileged usecase e.g. php server running on the pod network that has an arbitrary code execution vulnerability. It should be possible to limit further privilege escalation. The pod should not be able to:

1. exec into other containers
1. access a service account token
1. access the kube-env and various other sensitive stuff from the metadata server

First two were finished for 1.6. This solves (3) and gives us an extension point. A container proxy running on the host network with CAP_NET_ADMIN can implement any arbitrary policy and insert itself with:

```
iptables -t nat -I PREROUTING -p tcp -d 169.254.169.254 --dport 80 -j DNAT --to-destination 127.0.0.1:<PROXY_PORT>
iptables -t filter -I KUBE-METADATASERVER -j ACCEPT
```
We are not planning on building anything like this for 1.6, but might document how to enable select paths.

**Comment by mikedanese** [comment]
Also this will be opt in.

@k8s-bot kops aws e2e test this

**Comment by mikedanese** [comment]
@k8s-bot kops aws e2e test this

**Comment by justinsb** [comment]
@mikedanese AWS is on holiday today.

**Comment by mikedanese** [comment]
:sob:

**Comment by thockin** [comment]
On Tue, Feb 28, 2017 at 9:03 AM, Justin Santa Barbara
<notifications@github.com> wrote:
>
> Yes, for the container-aware proxy there are a few solutions, and kube2iam is widely used on kuberentes with AWS
>
> The challenge I think is figuring out iptables rules (or similar) that:
>
> don't let hostNetwork pods bypass it

hostNetwork pods are effectively indistinguishable from any other
daemon on the machine.  We either block or allow them all together.
Things like cgroup-based iptables rules are too new to rely on.

> let us give different permissions to k-c-m vs apiserver

If they are both hostNetwork, this is not viable.

> still allow us to run the container proxy kubernetes-managed, while letting it reach the IAM service and not blocking bootstrapping

I had assumed that hostNetwork is a privilege, and allowing that means
you allow access to metadata.  If that is not true, iptables is the
wrong approach.

**Comment by mikedanese** [comment]
@k8s-bot kops aws e2e test this

**Comment by justinsb** [comment]
@mikedanese so I put up a proposal which takes this and allows integration with something like kube2iam.  I am not sure if that is how you intended this to be used, or if you had something else in mind?  https://github.com/kubernetes/community/pull/501

**Pull request merged**